File: java_src/src/main/java/com/google/crypto/tink/hybrid/internal/HpkeContext.java
// Copyright 2021 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
///////////////////////////////////////////////////////////////////////////////

package com.google.crypto.tink.hybrid.internal;

import com.google.crypto.tink.proto.HpkePrivateKey;
import com.google.crypto.tink.proto.HpkePublicKey;
import com.google.crypto.tink.subtle.Bytes;
import com.google.crypto.tink.subtle.SubtleUtil;
import java.math.BigInteger;
import java.security.GeneralSecurityException;
import javax.annotation.concurrent.GuardedBy;
import javax.annotation.concurrent.ThreadSafe;

/**
 * Hybrid Public Key Encryption (HPKE) context for either a sender or a recipient.
 *
 * <p>https://www.ietf.org/archive/id/draft-irtf-cfrg-hpke-12.html#name-creating-the-encryption-con
 */
@ThreadSafe
final class HpkeContext {
  private static final byte[] EMPTY_IKM = new byte[0];

  private final HpkeAead aead;
  private final BigInteger maxSequenceNumber;
  private final byte[] key;
  private final byte[] baseNonce;
  private final byte[] encapsulatedKey;

  @GuardedBy("this")
  private BigInteger sequenceNumber;

  private HpkeContext(
      byte[] encapsulatedKey,
      byte[] key,
      byte[] baseNonce,
      BigInteger maxSequenceNumber,
      HpkeAead aead) {
    this.encapsulatedKey = encapsulatedKey;
    this.key = key;
    this.baseNonce = baseNonce;
    this.sequenceNumber = BigInteger.ZERO;
    this.maxSequenceNumber = maxSequenceNumber;
    this.aead = aead;
  }

  /** Helper function factored out to facilitate unit testing. */
  static HpkeContext createContext(
      byte[] encapsulatedKey,
      byte[] sharedSecret,
      HpkeKem kem,
      HpkeKdf kdf,
      HpkeAead aead,
      byte[] info)
      throws GeneralSecurityException {
    byte[] suiteId = HpkeUtil.hpkeSuiteId(kem.getKemId(), kdf.getKdfId(), aead.getAeadId());
    byte[] pskIdHash = kdf.labeledExtract(HpkeUtil.EMPTY_SALT, EMPTY_IKM, "psk_id_hash", suiteId);
    byte[] infoHash = kdf.labeledExtract(HpkeUtil.EMPTY_SALT, info, "info_hash", suiteId);
    byte[] keyScheduleContext = Bytes.concat(HpkeUtil.BASE_MODE, pskIdHash, infoHash);
    byte[] secret = kdf.labeledExtract(sharedSecret, EMPTY_IKM, "secret", suiteId);

    byte[] key = kdf.labeledExpand(secret, keyScheduleContext, "key", suiteId, aead.getKeyLength());
    byte[] baseNonce =
        kdf.labeledExpand(secret, keyScheduleContext, "base_nonce", suiteId, aead.getNonceLength());
    BigInteger maxSeqNo = maxSequenceNumber(aead.getNonceLength());

    return new HpkeContext(encapsulatedKey, key, baseNonce, maxSeqNo, aead);
  }

  /**
   * Creates HPKE sender context according to KeySchedule() defined in
   * https://www.ietf.org/archive/id/draft-irtf-cfrg-hpke-12.html#section-5.1-9.
   *
   * @param recipientPublicKey recipient's public key (pkR)
   * @param kem key encapsulation mechanism primitive
   * @param kdf key derivation function primitive
   * @param aead authenticated encryption with associated data primitive
   * @param info application-specific information parameter to influence key generation
   */
  static HpkeContext createSenderContext(
      HpkePublicKey recipientPublicKey, HpkeKem kem, HpkeKdf kdf, HpkeAead aead, byte[] info)
      throws GeneralSecurityException {
    HpkeKemEncapOutput encapOutput =
        kem.encapsulate(recipientPublicKey.getPublicKey().toByteArray());
    byte[] encapsulatedKey = encapOutput.getEncapsulatedKey();
    byte[] sharedSecret = encapOutput.getSharedSecret();
    return createContext(encapsulatedKey, sharedSecret, kem, kdf, aead, info);
  }

  /**
   * Creates HPKE sender recipient context according to KeySchedule() defined in
   * https://www.ietf.org/archive/id/draft-irtf-cfrg-hpke-12.html#section-5.1-9.
   *
   * @param encapsulatedKey encapsulated key (enc)
   * @param recipientPrivateKey recipient's private key (skR)
   * @param kem key encapsulation mechanism primitive
   * @param kdf key derivation function primitive
   * @param aead authenticated encryption with associated data primitive
   * @param info application-specific information parameter to influence key generation
   */
  static HpkeContext createRecipientContext(
      byte[] encapsulatedKey,
      HpkePrivateKey recipientPrivateKey,
      HpkeKem kem,
      HpkeKdf kdf,
      HpkeAead aead,
      byte[] info)
      throws GeneralSecurityException {
    byte[] sharedSecret =
        kem.decapsulate(encapsulatedKey, recipientPrivateKey.getPrivateKey().toByteArray());
    return createContext(encapsulatedKey, sharedSecret, kem, kdf, aead, info);
  }

  private static BigInteger maxSequenceNumber(int nonceLength) {
    return BigInteger.ONE.shiftLeft(8 * nonceLength).subtract(BigInteger.ONE);
  }

  @GuardedBy("this")
  private void incrementSequenceNumber() throws GeneralSecurityException {
    if (sequenceNumber.compareTo(maxSequenceNumber) >= 0) {
      throw new GeneralSecurityException("message limit reached");
    }
    sequenceNumber = sequenceNumber.add(BigInteger.ONE);
  }

  /**
   * ComputeNonce() from
   * https://www.ietf.org/archive/id/draft-irtf-cfrg-hpke-12.html#section-5.2-11.
   */
  @GuardedBy("this")
  private byte[] computeNonce() throws GeneralSecurityException {
    return Bytes.xor(baseNonce, SubtleUtil.integer2Bytes(sequenceNumber, aead.getNonceLength()));
  }

  /** Returns the next nonce to use for seal/open. Also, increments the sequence number. */
  private synchronized byte[] computeNonceAndIncrementSequenceNumber()
      throws GeneralSecurityException {
    byte[] nonce = computeNonce();
    incrementSequenceNumber();
    return nonce;
  }

  byte[] getKey() {
    return key;
  }

  byte[] getBaseNonce() {
    return baseNonce;
  }

  byte[] getEncapsulatedKey() {
    return encapsulatedKey;
  }

  /**
   * Performs AEAD encryption of {@code plaintext} with {@code associatedData} according to
   * ContextS.Seal() defined in
   * https://www.ietf.org/archive/id/draft-irtf-cfrg-hpke-12.html#section-5.2-8.
   *
   * @return ciphertext
   */
  byte[] seal(byte[] plaintext, byte[] associatedData) throws GeneralSecurityException {
    byte[] nonce = computeNonceAndIncrementSequenceNumber();
    return aead.seal(key, nonce, plaintext, associatedData);
  }

  /**
   * Performs AEAD decryption of {@code ciphertext} with {@code associatedData} according to
   * ContextR.Open() defined in
   * https://www.ietf.org/archive/id/draft-irtf-cfrg-hpke-12.html#section-5.2-10.
   *
   * @return plaintext
   */
  byte[] open(byte[] ciphertext, byte[] associatedData) throws GeneralSecurityException {
    byte[] nonce = computeNonceAndIncrementSequenceNumber();
    return aead.open(key, nonce, ciphertext, associatedData);
  }
}

File: javascript/hybrid/internal/hpke/hpke_context.ts
/**
 * @license
 * Copyright 2022 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */

import {SecurityException} from '../../../exception/security_exception';
import * as bytes from '../../../subtle/bytes';

import {HpkeAead} from './hpke_aead';
import {HpkeKdf} from './hpke_kdf';
import {HpkeKem} from './hpke_kem';
import {HpkeKemEncapOutput} from './hpke_kem_encap_output';
import * as hpkeUtil from './hpke_util';
import {NistCurvesHpkeKemPrivateKey} from './nist_curves_hpke_kem_private_key';

/**
 * Hybrid Public Key Encryption (HPKE) context for either a sender or a
 * recipient.
 *
 * @see https://www.rfc-editor.org/rfc/rfc9180.html#name-creating-the-encryption-con
 */
export class HpkeContext {
  private sequenceNumber: bigint;
  private readonly maxSequenceNumber: bigint;

  constructor(
      private readonly encapsulatedKey: Uint8Array,
      private readonly key: Uint8Array, private readonly baseNonce: Uint8Array,
      private readonly aead: HpkeAead) {
    this.sequenceNumber = BigInt(0);

    /**
     * Indicates that the message limit is reached, calculated as per
     * @see https://www.rfc-editor.org/rfc/rfc9180.html#section-5.2-11.
     */
    this.maxSequenceNumber =
        (BigInt(1) << BigInt(8 * this.aead.getNonceLength())) - BigInt(1);
  }

  /**
   * Performs AEAD encryption of `plaintext` with `associatedData`
   * according to `ContextS.Seal()` as defined in
   * @see https://www.rfc-editor.org/rfc/rfc9180.html#section-5.2-8.
   */
  async seal(plaintext: Uint8Array, associatedData: Uint8Array):
      Promise<Uint8Array> {
    const nonce: Uint8Array = this.computeNonceAndIncrementSequenceNumber();
    return await this.aead.seal(
        {key: this.key, nonce, plaintext, associatedData});
  }

  /**
   * Performs AEAD decryption of `ciphertext` with `associatedData`
   * according to `ContextR.Open()` as defined in
   * @see https://www.rfc-editor.org/rfc/rfc9180.html#section-5.2-10.
   */
  async open(ciphertext: Uint8Array, associatedData: Uint8Array):
      Promise<Uint8Array> {
    const nonce: Uint8Array = this.computeNonceAndIncrementSequenceNumber();
    return this.aead.open({key: this.key, nonce, ciphertext, associatedData});
  }

  /**
   * Computes nonce according to `ComputeNonce` as defined in
   * @see https://www.rfc-editor.org/rfc/rfc9180.html#section-5.2-11.
   * for seal/open and increments the sequence number.
   */
  private computeNonceAndIncrementSequenceNumber(): Uint8Array {
    const seqBytes: Uint8Array = hpkeUtil.bigIntToByteArray(
        this.aead.getNonceLength(), this.sequenceNumber);
    const nonce: Uint8Array = bytes.xor(this.baseNonce, seqBytes);

    if (this.sequenceNumber >= this.maxSequenceNumber) {
      throw new SecurityException('message limit reached');
    }
    this.sequenceNumber += (BigInt(1));

    return nonce;
  }

  getKey(): Uint8Array {
    return this.key;
  }

  getBaseNonce(): Uint8Array {
    return this.baseNonce;
  }

  getEncapsulatedKey(): Uint8Array {
    return this.encapsulatedKey;
  }
}

/** Helper function factored out to facilitate unit testing. */
export async function createContext(
    encapsulatedKey: Uint8Array, sharedSecret: Uint8Array, kem: HpkeKem,
    kdf: HpkeKdf, aead: HpkeAead, info: Uint8Array): Promise<HpkeContext> {
  const suiteId = hpkeUtil.hpkeSuiteId(
      {kemId: kem.getKemId(), kdfId: kdf.getKdfId(), aeadId: aead.getAeadId()});

  /**
   * The IKM values below are empty because we only support base mode
   * @see https://www.rfc-editor.org/rfc/rfc9180.html#section-5.1.9.
   */
  const pskIdHash = await kdf.labeledExtract({
    ikm: new Uint8Array(0),  // empty IKM
    ikmLabel: 'psk_id_hash',
    suiteId,
  });

  const infoHash = await kdf.labeledExtract({
    ikm: info,
    ikmLabel: 'info_hash',
    suiteId,
  });

  const keyScheduleContext =
      bytes.concat(hpkeUtil.BASE_MODE, pskIdHash, infoHash);
  const secret = await kdf.labeledExtract({
    ikm: new Uint8Array(0),  // empty IKM
    ikmLabel: 'secret',
    suiteId,
    salt: sharedSecret
  });

  const key = await kdf.labeledExpand({
    prk: secret,
    info: keyScheduleContext,
    infoLabel: 'key',
    suiteId,
    length: aead.getKeyLength()
  });
  const baseNonce = await kdf.labeledExpand({
    prk: secret,
    info: keyScheduleContext,
    infoLabel: 'base_nonce',
    suiteId,
    length: aead.getNonceLength()
  });

  return new HpkeContext(encapsulatedKey, key, baseNonce, aead);
}

/**
 * Creates HPKE sender context according to `KeySchedule` as defined in
 * @see https://www.rfc-editor.org/rfc/rfc9180.html#section-5.1-9.
 */
export async function createSenderContext(
    recipientPublicKey: Uint8Array, kem: HpkeKem, kdf: HpkeKdf, aead: HpkeAead,
    info: Uint8Array): Promise<HpkeContext> {
  const encapOutput: HpkeKemEncapOutput =
      await kem.encapsulate(recipientPublicKey);
  const encapsulatedKey: Uint8Array = encapOutput.encapsulatedKey;
  const sharedSecret: Uint8Array = encapOutput.sharedSecret;
  return await createContext(
      encapsulatedKey, sharedSecret, kem, kdf, aead, info);
}

/**
 * Creates HPKE recipient context according to `KeySchedule` as defined in
 * @see https://www.rfc-editor.org/rfc/rfc9180.html#section-5.1-9.
 */
export async function createRecipientContext(
    encapsulatedKey: Uint8Array,
    recipientPrivateKey: NistCurvesHpkeKemPrivateKey, kem: HpkeKem,
    kdf: HpkeKdf, aead: HpkeAead, info: Uint8Array): Promise<HpkeContext> {
  const sharedSecret: Uint8Array =
      await kem.decapsulate(encapsulatedKey, recipientPrivateKey);
  return await createContext(
      encapsulatedKey, sharedSecret, kem, kdf, aead, info);
}
